src/Framework/Framework/Compilation/Binding/GeneralBindingPropertyResolvers.cs

```diff
@@ -53,12 +54,18 @@ public ActionFiltersBindingProperty GetActionFilters(ParsedExpressionBindingProp
         }
 
         public Expression<BindingDelegate> CompileToDelegate(
-            CastedExpressionBindingProperty expression, DataContextStack? dataContext = null)
+            CastedExpressionBindingProperty expression, IBinding binding, DataContextStack? dataContext = null)
         {
-            var expr = BindingCompiler.ReplaceParameters(expression.Expression, dataContext);
+            var replacementVisitor = new BindingCompiler.ParameterReplacementVisitor(dataContext);
+            var expr = replacementVisitor.Visit(expression.Expression);
             expr = new ExpressionNullPropagationVisitor(e => true).Visit(expr);
             expr = ExpressionUtils.ConvertToObject(expr);
-            return Expression.Lambda<BindingDelegate>(expr, BindingCompiler.ViewModelsParameter, BindingCompiler.CurrentControlParameter);
+            expr = replacementVisitor.WrapExpression(expr, contextObject: binding);
+            var l = Expression.Lambda<BindingDelegate>(expr, BindingCompiler.CurrentControlParameter);
```

Comment:
I renamed all other `l` variables in our codebase (I wasn't lazy only here :] ). I'll leave the commented `Console.WriteLine`s here, I think it will save us some searching next time there is a bug the BindingCompiler, NullPropagationVisitor or something like it.